### PR TITLE
(Fix #341) SYSENTER instruction is not properly hooked with uc_hook_add in x86 emulation

### LINKED ISSUE #341: SYSENTER instruction is not properly hooked in x86 emulation.
[helper_sysenter](https://github.com/unicorn-engine/unicorn/blob/ed319bda0b7f7b6bb4a946d24682897e7f465c23/qemu/target-i386/seg_helper.c#L2304) don't check properly if a call interrupt callback is registred with uc_hook_add.

### PR DESCRIPTION
**Pull request by Spl3en**
helper_sysenter in qemu/target-i386/seg_helper.c didn't check properly if a call interrupt callback was registred.
It has been fixed by copying the helper_syscall behavior.

### CODE REVIEW

File: qemu/target-i386/translate.c [hunk 1]
```diff
@@ -7425,14 +7425,11 @@ static target_ulong disas_insn(CPUX86State *env, DisasContext *s,
         /* For Intel SYSENTER is valid on 64-bit */
         if (CODE64(s) && env->cpuid_vendor1 != CPUID_VENDOR_INTEL_1)
             goto illegal_op;
-        if (!s->pe) {
-            gen_exception(s, EXCP0D_GPF, pc_start - s->cs_base);
-        } else {
```
Comment (aquynh) [inline qemu/target-i386/translate.c]: why did you remove this check on `!s->pe`?

Comment (Spl3en) [inline qemu/target-i386/translate.c]: I admit it is purely based on my observation on how SYSCALL is handled. Also, during my tests, this condition was true, so there might be another issue explaining why sysenter_helper is not called.

Comment (aquynh) [inline qemu/target-i386/translate.c]: if you recover this check (that you removed), is your testcase still working?

**Comment by aquynh** [comment]
also, please dont forget to add `sysenter_hook_x86` to `tests/regress/Makefile` to compile it, and add `sysenter_hook_x86` to `.gitignore`.

File: qemu/target-i386/translate.c (hunk 1 above)
Comment (Spl3en) [inline qemu/target-i386/translate.c]: If I recover this check it doesn't work, the sysenter hook is not called anymore.

**Comment by Spl3en** [comment]
Done.

**Comment by aquynh** [comment]
that branch should not be removed. instead, we need to initialize the engine properly in protected mode. currently the CPU is not in protected mode, hence your observation.

**Comment by Spl3en** [comment]
Alright, I was sure what `pe` meant, I understand better now. I've restored the check.

**Comment by aquynh** [comment]
will be great if you can find how to set engine in protected mode when the engine is created. (hint: the flag HF_PE_SHIFT)

**Comment by Spl3en** [comment]
I think I found the reason why it failed to be in protected mode.

CR0 was not correctly reset in [unicorn.c : x86_reg_reset](https://github.com/unicorn-engine/unicorn/blob/51323c9c17a32f204b169bebc10d1c79553e966f/qemu/target-i386/unicorn.c#L135).
We need to call [cpu_x86_update_cr0](https://github.com/Spl3en/unicorn/commit/bb375e4fa9d68bf2878048d77ef07db7e209f1b1#diff-92173f1ce0d5319bbb03193432e9205dR135) instead of manually setting CR0, so the hflags is also updated correctly.

Now the test passes correctly.

**Comment by farmdve** [comment]
@aquynh 

Do you think we should design another API, for instance called emu_opt_set e.g emu options set where we can specify processor-specific things?

E.g protected mode, long mode, real mode etc. Maybe even other interesting flags.

**Comment by lunixbochs** [comment]
I second this, but I don't think it needs the "emu" prefix. Also kinda related to https://github.com/unicorn-engine/unicorn/tree/cpu_model

**Comment by farmdve** [comment]
@lunixbochs 

I'm just trying to stick to the standards. Not having the emu part would make it look weird compared to the rest of the API.

**Comment by lunixbochs** [comment]
You already have a valid prefix (opt). uc_emu_opt_set is just long-winded.

**Comment by farmdve** [comment]
Ah, brain fart on my part then. I honestly forgot about the uc prefix. uc_opt_set then.

**Comment by aquynh** [comment]
yes i proposed a new API for customize all kind of runtime options, like `uc_option()`. this will be similar to `cs_option()` API of Capstone: https://github.com/aquynh/capstone/blob/master/include/capstone.h#L383

**Comment by aquynh** [comment]
merged, thanks!
